javascript/ql/src/semmle/javascript/StandardLibrary.qll

```diff
@@ -224,3 +224,27 @@ private class PromiseTaintStep extends TaintTracking::AdditionalTaintStep {
     pred = source and succ = this
   }
 }
+
+/**
+ * A flow step propagating the exception thrown from a callback to a method whose name coincides
+ * a built-in Array iteration method, such as `forEach` or `map`.
+ */
+private class IteratorExceptionStep extends DataFlow::MethodCallNode, DataFlow::AdditionalFlowStep {
+  IteratorExceptionStep() {
+    exists(string name | name = getMethodName() |
+      name = "forEach" or
+      name = "each" or
+      name = "map" or
+      name = "filter" or
+      name = "some" or
+      name = "every" or
+      name = "fold" or
+      name = "reduce"
+    )
+  }
+
+  override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
+    pred = getAnArgument().(DataFlow::FunctionNode).getExceptionalReturn() and
```

Comment:
Use `getCallback(_)` for a bit of extra local data flow? Probably won't make a difference in practice, but might as well.

Reply:
It's the same story as for taint steps out of `map` callbacks: if the function has more than one use, it can lead to mismatched call/return steps. We're adding a *local* step that technically crosses function boundaries.

Reply:
Right; makes sense.